hivemind/p2p/p2p_daemon_bindings/control.py

```diff
@@ -90,8 +104,135 @@ async def listen(self) -> AsyncIterator["ControlClient"]:
+
+        await self._pending_messages.put(
+            p2pd_pb.PersistentConnectionRequest(
+                callId=call_id.bytes,
+                unaryResponse=response,
+            )
+        )
+        self._handler_tasks.pop(call_id)
+
+    async def _handler(self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter):
+        pb_stream_info = p2pd_pb.StreamInfo()  # type: ignore
+        await read_pbmsg_safe(reader, pb_stream_info)
+        stream_info = StreamInfo.from_protobuf(pb_stream_info)
+        try:
+            handler = self.handlers[stream_info.proto]
+        except KeyError as e:
+            # should never enter here... daemon should reject the stream for us.
+            writer.close()
+            raise DispatchFailure(e)
+        await handler(stream_info, reader, writer)
+
+    async def _send_call_cancel(self, call_id: uuid.UUID):
+        await self._pending_messages.put(
+            p2pd_pb.PersistentConnectionRequest(
+                callId=call_id.bytes,
+                cancel=p2pd_pb.Cancel(),
+            ),
+        )
+
+    async def _ensure_persistent_conn(self):
+        if not self._is_persistent_conn_open:
+            async with self._ensure_conn_lock:
+                if not self._is_persistent_conn_open:
+                    reader, writer = await self.daemon_connector.open_persistent_connection()
+
+                    self._read_task = asyncio.create_task(self._read_from_persistent_conn(reader))
+                    self._write_task = asyncio.create_task(self._write_to_persistent_conn(writer))
+
+                    self._is_persistent_conn_open = True
+
+    async def add_unary_handler(self, proto: str, handler: TUnaryHandler):
+        await self._ensure_persistent_conn()
+
+        call_id = uuid.uuid4()
+
+        add_unary_handler_req = p2pd_pb.AddUnaryHandlerRequest(proto=proto)
+        req = p2pd_pb.PersistentConnectionRequest(callId=call_id.bytes, addUnaryHandler=add_unary_handler_req)
+
+        if self.unary_handlers.get(proto):
+            raise ValueError(f"Handler for protocol {proto} already assigned")
+        self.unary_handlers[proto] = handler
+
+        await self._pending_messages.put(req)
+
+    async def call_unary_handler(self, peer_id: PeerID, proto: str, data: bytes) -> bytes:
+        call_id = uuid.uuid4()
+        call_unary_req = p2pd_pb.CallUnaryRequest(
+            peer=peer_id.to_bytes(),
+            proto=proto,
+            data=data,
+        )
+        req = p2pd_pb.PersistentConnectionRequest(
+            callId=call_id.bytes,
+            callUnary=call_unary_req,
+        )
+
+        await self._ensure_persistent_conn()
+
+        try:
+            self._pending_calls[call_id] = asyncio.Future()
+            await self._pending_messages.put(req)
+            return await self._pending_calls[call_id]
+
+        except asyncio.CancelledError:
+            asyncio.create_task(self._send_call_cancel(call_id))
```

Comment:
```suggestion
            await self._send_call_cancel(call_id)
```

I suggest to stick with `await` unless it creates more warnings.

The reason is that if `self._send_call_cancel` fails, the error won't be hidden and will be propagated to the caller.

I remember that I suggested to try `asyncio.create_task()` earlier, but that was an attempt to deal with the asyncio warnings (that were fixed in #346).